src/ring_theory/graded_algebra/radical.lean

```diff
@@ -0,0 +1,184 @@
+  have mem_max₂ : max₂ ∈ set₂ := max'_mem set₂ (nonempty y rid₂),
+  replace hxy : proj 𝒜 (max₁ + max₂) (x * y) ∈ I := hI _ hxy,
+
+  have mem_I : proj 𝒜 max₁ x * proj 𝒜 max₂ y ∈ I,
+  { set antidiag :=
+      ((support 𝒜 x).product (support 𝒜 y)).filter (λ z : ι × ι, z.1 + z.2 = max₁ + max₂) with ha,
+    have mem_antidiag : (max₁, max₂) ∈ antidiag,
+    { simp only [add_sum_erase, mem_filter, mem_product],
+      exact ⟨⟨mem_of_mem_filter _ mem_max₁, mem_of_mem_filter _ mem_max₂⟩, rfl⟩ },
+    have eq_add_sum :=
+      calc  proj 𝒜 (max₁ + max₂) (x * y)
+          = ∑ ij in antidiag, proj 𝒜 ij.1 x * proj 𝒜 ij.2 y
+          : by simp_rw [ha, proj_apply, map_mul, support, direct_sum.coe_mul_apply_submodule]
+      ... = proj 𝒜 max₁ x * proj 𝒜 max₂ y + ∑ ij in antidiag.erase (max₁, max₂),
+                                              proj 𝒜 ij.1 x * proj 𝒜 ij.2 y
+          : (add_sum_erase _ _ mem_antidiag).symm,
+    rw eq_sub_of_add_eq eq_add_sum.symm,
+    refine ideal.sub_mem _ hxy (ideal.sum_mem _ (λ z H, _)),
+    rcases z with ⟨i, j⟩,
+    simp only [mem_erase, prod.mk.inj_iff, ne.def, mem_filter, mem_product] at H,
+    rcases H with ⟨H₁, ⟨H₂, H₃⟩, H₄⟩,
+    have max_lt : max₁ < i ∨ max₂ < j,
+    { rcases lt_trichotomy max₁ i with h | rfl | h,
+      { exact or.inl h },
+      { refine false.elim (H₁ ⟨rfl, add_left_cancel H₄⟩), },
+      { apply or.inr,
+        have := add_lt_add_right h j,
+        rw H₄ at this,
+        exact lt_of_add_lt_add_left this, }, },
+    cases max_lt,
+    { -- in this case `max₁ < i`, then `xᵢ ∈ I`; for otherwise `i ∈ set₁` then `i ≤ max₁`.
+      have not_mem : i ∉ set₁ := λ h, lt_irrefl _
+        ((max'_lt_iff set₁ (nonempty x rid₁)).mp max_lt i h),
+      rw set₁_eq at not_mem,
+      simp only [not_and, not_not, ne.def, dfinsupp.mem_support_to_fun, mem_filter] at not_mem,
+      exact ideal.mul_mem_right _ I (not_mem H₂), },
+    { -- in this case  `max₂ < j`, then `yⱼ ∈ I`; for otherwise `j ∈ set₂`, then `j ≤ max₂`.
+      have not_mem : j ∉ set₂ := λ h, lt_irrefl _
+        ((max'_lt_iff set₂ (nonempty y rid₂)).mp max_lt j h),
+      rw set₂_eq at not_mem,
+      simp only [not_and, not_not, ne.def, dfinsupp.mem_support_to_fun, mem_filter] at not_mem,
+      exact ideal.mul_mem_left I _ (not_mem H₃), }, },
+
+  have not_mem_I : proj 𝒜 max₁ x * proj 𝒜 max₂ y ∉ I,
+  { have neither_mem : proj 𝒜 max₁ x ∉ I ∧ proj 𝒜 max₂ y ∉ I,
+    { rw mem_filter at mem_max₁ mem_max₂,
+      exact ⟨mem_max₁.2, mem_max₂.2⟩, },
+    intro rid,
+    cases homogeneous_mem_or_mem ⟨max₁, submodule.coe_mem _⟩ ⟨max₂, submodule.coe_mem _⟩ mem_I,
+    { apply neither_mem.1 h },
+    { apply neither_mem.2 h }, },
+
+  exact not_mem_I mem_I,
+end⟩
+
+lemma homogeneous_ideal.is_prime_iff (I : homogeneous_ideal 𝒜) :
+  I.1.is_prime ↔
+  (I ≠ ⊤) ∧
+    ∀ {x y : A}, set_like.is_homogeneous 𝒜 x → set_like.is_homogeneous 𝒜 y
+      → (x * y ∈ I.1 → x ∈ I.1 ∨ y ∈ I.1) :=
```

Comment:
I think this statement (and maybe all the statements in this file) would be more useful unbundled:
```suggestion
lemma ideal.is_homogeneous.is_prime_iff {I : ideal A} (h : ideal.is_homogeneous 𝒜) :
  I.is_prime ↔
  (I ≠ ⊤) ∧
    ∀ {x y : A}, set_like.is_homogeneous 𝒜 x → set_like.is_homogeneous 𝒜 y
      → (x * y ∈ I → x ∈ I ∨ y ∈ I) :=
```
If nothing else, `.1` is not simp normal form, `coe` is.

Feel free to keep the bundled statements as well if you want them, which are trivial to prove in terms of the unbundled ones.